streams/src/main/java/org/apache/kafka/streams/state/internals/VersionedKeyValueToBytesStoreAdapter.java

```diff
@@ -0,0 +1,196 @@
+
+    public VersionedKeyValueToBytesStoreAdapter(final VersionedKeyValueStore<Bytes, byte[]> inner) {
+        this.inner = Objects.requireNonNull(inner);
+    }
+
+    @Override
+    public void put(final Bytes key, final byte[] rawValueAndTimestamp) {
+        if (rawValueAndTimestamp == null) {
+            throw new IllegalArgumentException("Put to versioned store must always include timestamp, including for tombstones.");
+        }
+        final ValueAndTimestamp<byte[]> valueAndTimestamp
+            = VALUE_AND_TIMESTAMP_DESERIALIZER.deserialize(null, rawValueAndTimestamp);
+        inner.put(
+            key,
+            valueAndTimestamp.value(),
+            valueAndTimestamp.timestamp()
+        );
+    }
+
+    @Override
+    public byte[] get(final Bytes key) {
+        final VersionedRecord<byte[]> versionedRecord = inner.get(key);
+        return serializeAsBytes(versionedRecord);
+    }
+
+    @Override
+    public byte[] get(final Bytes key, final long asOfTimestamp) {
+        final VersionedRecord<byte[]> versionedRecord = inner.get(key, asOfTimestamp);
+        return serializeAsBytes(versionedRecord);
+    }
+
+    @Override
+    public byte[] delete(final Bytes key, final long timestamp) {
+        final VersionedRecord<byte[]> versionedRecord = inner.delete(key, timestamp);
+        return serializeAsBytes(versionedRecord);
+    }
+
+    @Override
+    public String name() {
+        return inner.name();
+    }
+
+    @Deprecated
+    @Override
+    public void init(final ProcessorContext context, final StateStore root) {
+        inner.init(context, root);
+    }
+
+    @Override
+    public void init(final StateStoreContext context, final StateStore root) {
+        inner.init(context, root);
+    }
+
+    @Override
+    public void flush() {
+        inner.flush();
+    }
+
+    @Override
+    public void close() {
+        inner.close();
+    }
+
+    @Override
+    public boolean persistent() {
+        return inner.persistent();
+    }
+
+    @Override
+    public boolean isOpen() {
+        return inner.persistent();
+    }
+
+    @Override
+    public <R> QueryResult<R> query(final Query<R> query, final PositionBound positionBound, final QueryConfig config) {
+        return inner.query(query, positionBound, config);
+    }
+
+    @Override
+    public Position getPosition() {
+        return inner.getPosition();
+    }
+
+    @Override
+    public byte[] putIfAbsent(final Bytes key, final byte[] value) {
+        throw new UnsupportedOperationException("Versioned key-value stores do not support putIfAbsent(key, value)");
+    }
+
+    @Override
+    public void putAll(final List<KeyValue<Bytes, byte[]>> entries) {
+        throw new UnsupportedOperationException("Versioned key-value stores do not support putAll(entries)");
+    }
+
+    @Override
+    public byte[] delete(final Bytes key) {
+        throw new UnsupportedOperationException("Versioned key-value stores do not support delete(key)");
```

Comment:
Should we point to `delete(key, ts)` in the error message?

Reply:
Sure, updated.